vizro-ai/src/vizro_ai/dashboard/nodes/core_builder/build.py

```diff
@@ -0,0 +1,93 @@
+"""Module that contains the builder functionality."""
+
+import logging
+
+import vizro.models as vm
+from vizro_ai.utils.helper import DebugFailure
+
+logger = logging.getLogger(__name__)
+logger.setLevel(logging.INFO)
+
+
+class PageBuilder:
+    """Class to build a page."""
+
+    def __init__(self, model, df_metadata, page_plan):
+        """Initialize PageBuilder."""
+        self._model = model
+        self._df_metadata = df_metadata
+        self._page_plan = page_plan
+        self._components = None
+        self._controls = None
+        self._page = None
+        self._layout = None
+
+    @property
+    def components(self):
+        """Property to get components."""
+        if self._components is None:
+            self._components = self._build_components()
+        return self._components
+
+    def _build_components(self):
+        components = []
+        logger.info(f"Building components of page: {self._page_plan.title}")
+        # Could potentially be parallelized or sent as a batch to the API
+        for i in range(len(self._page_plan.components)):
+            logger.info(f"{self._page_plan.title} -> Building component {self._page_plan.components[i]}")
+            try:
+                components.append(
+                    self._page_plan.components[i].create(df_metadata=self._df_metadata, model=self._model)
+                )
+            except DebugFailure as e:  # TODO: check - does this ever get raised?
+                components.append(
+                    vm.Card(id=self._page_plan.components[i].component_id, text=f"Failed to build component: {e}")
+                )
+        return components
+
+    @property
+    def layout(self):
+        """Property to get layout."""
+        if self._layout is None:
+            self._layout = self._build_layout()
+        return self._layout
+
+    def _build_layout(self):
```

Comment:
tbd `grid-template-area`.

Reply:
implemented this logic to layout creation and did several tests so far. Seems like the human language -> grid-template-area is not always reliable. The second step (grid-template-area -> Vizro layout grid) is quite stable and correct.